### PR TITLE
Guard second doc parsing pass with index setting

### PR DESCRIPTION
**Pull request by kkrik-es**
This may prove handy in case we see issues like https://github.com/elastic/elasticsearch/issues/114563, as a tentative workaround.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java [hunk 1]
```diff
@@ -371,7 +371,7 @@ public final Collection<IgnoredSourceFieldMapper.NameValue> getIgnoredFieldsMiss
      * Applies to synthetic source only.
      */
     public final DocumentParserContext maybeCloneForArray(Mapper mapper) throws IOException {
-        if (canAddIgnoredField() && mapper instanceof ObjectMapper) {
+        if (canAddIgnoredField() && mapper instanceof ObjectMapper && indexSettings().isSyntheticSourceSecondDocParsingPassEnabled()) {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:374]: Can we avoid calling this (reading a volatile field) per `array` field and do it once per document instead? I'm benchmarking the impact of https://github.com/elastic/elasticsearch/blob/363143fa39822300534a8d4f80aa267c6b2ede59/server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java#L444.

**Review by dnhatn — APPROVED** [review]
One comment, but LGTM. Thanks Kostas!

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java (hunk 1 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java]: Interesting, I didn't realize that volatiles are so expensive.. Done.

**Review by lkts — APPROVED** [review]

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by kkrik-es** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/114799)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**